src/api/app/services/scm_status_reporter.rb

```diff
@@ -17,47 +19,25 @@ def initialize(event_payload, event_subscription_payload, scm_token, workflow_ru
   end
 
   def call
-    if github?
-      GithubStatusReporter.new(@event_payload,
-                               @event_subscription_payload,
-                               @scm_token,
-                               @state,
-                               @workflow_run,
-                               @event_type,
-                               initial_report: @initial_report).call
-    elsif gitlab?
-      GitlabStatusReporter.new(@event_payload,
-                               @event_subscription_payload,
-                               @scm_token,
-                               @state,
-                               @workflow_run,
-                               @event_type,
-                               initial_report: @initial_report).call
-    elsif gitea?
-      GiteaStatusReporter.new(@event_payload,
+    return if EVENT_ACTIONS_TO_SKIP.include?(@event_payload[:action])
+
+    detect_reporter_class.new(@event_payload,
                               @event_subscription_payload,
                               @scm_token,
                               @state,
                               @workflow_run,
                               @event_type,
                               initial_report: @initial_report).call
-    end
-  end
-
-  def github?
-    @event_subscription_payload[:scm] == 'github'
   end
 
-  def gitlab?
-    @event_subscription_payload[:scm] == 'gitlab'
-  end
+  private
 
-  def gitea?
-    @event_subscription_payload[:scm] == 'gitea'
+  # This is safe to do as there are no ways to abuse this other than forging an
+  # SCM event payload AND inject a custom status reporter into our codebase
+  def detect_reporter_class
+    "#{@event_subscription_payload[:scm].capitalize}StatusReporter".constantize
```

Comment:
I tend to prefer to rely on an `key:value` map instead, where the name of the class is not a combination of strings and dynamic values, but they are defined in advance. I find it more readable and stable in case names of classes change or whatever else at some point changes too. But maybe it's just a personal preference.

You could even use it directly inline:
i.e.:
```
REPORTERS = {
    'github' => GithubStatusReporter,
    'gitlab' => GitlabStatusReporter,
    'gitea' => GiteaStatusReporter
  }.freeze

...
...

REPORTERS[@event_subscription_payload[:scm]].new(@event_payload,
                                                                                                  @event_subscription_payload,
                                                                                                  @scm_token,
...
...
```

Feel free to evaluate it or discard it.

Reply:
Alright, I'll use this instead of metaprograming. But I'll do it in #16422.